File: app/src/main/java/com/kina/care/presentation/profile/ProfileViewModel.kt
package com.kina.care.presentation.profile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.kina.care.domain.model.LeaveRequest
import com.kina.care.domain.model.User
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

data class ProfileUiState(
    val isLoading: Boolean = true,
    val userData: User? = null,
    val errorMessage: String? = null,
    val isSaving: Boolean = false,
    val isSaveSuccess: Boolean = false,
    val registrationFee: Double = 1000.0
)

class ProfileViewModel : ViewModel() {
    private val firestore = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()
    private val _uiState = MutableStateFlow(ProfileUiState())
    val uiState: StateFlow<ProfileUiState> = _uiState.asStateFlow()

    init {
        fetchUserProfile()
        fetchSystemSettings()
    }

    private fun fetchUserProfile() {
        val userId = auth.currentUser?.uid ?: return
        firestore.collection("users").document(userId)
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null && snapshot.exists()) {
                    val user = snapshot.toObject(User::class.java)
                    _uiState.value = _uiState.value.copy(isLoading = false, userData = user)

                    // 🌟 Fix: CAREGIVER/NURSE ලට Admin KYC approve කළ සැනිකිනිම isVerified = true set කිරීම
                    // CLIENT ලා registration fee ගෙව්වාට පස්සෙ payRegistrationFee() function එකෙ isVerified = true වෙනවා
                    val isCaregiverRole = user?.role == "CAREGIVER" || user?.role == "NURSE"
                    if (user != null && isCaregiverRole && user.kycStatus == "APPROVED" && !user.isVerified) {
                        firestore.collection("users").document(userId)
                            .update("isVerified", true)
                    }
                }
            }
    }

    private fun fetchSystemSettings() {
        firestore.collection("settings").document("general")
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null && snapshot.exists()) {
                    val fee = snapshot.getDouble("registrationFee") ?: 1000.0
                    _uiState.value = _uiState.value.copy(registrationFee = fee)
                }
            }
    }

    fun requestLeave(dateMs: Long, reason: String, onResult: (Boolean, String) -> Unit) {
        viewModelScope.launch {
            try {
                val userId = auth.currentUser?.uid ?: return@launch
                val userName = _uiState.value.userData?.name ?: "Caregiver"
                val dateStr = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date(dateMs))

                val leaveId = java.util.UUID.randomUUID().toString()
                val leaveReq = LeaveRequest(
                    id = leaveId,
                    caregiverId = userId,
                    caregiverName = userName,
                    dateMs = dateMs,
                    dateString = dateStr,
                    reason = reason,
                    status = "PENDING",
                    requestedAt = System.currentTimeMillis()
                )

                firestore.collection("leave_requests").document(leaveId).set(leaveReq).await()
                onResult(true, "නිවාඩු ඉල්ලීම යවන ලදී. Admin අනුමැතිය ලැබෙන තෙක් රැඳී සිටින්න.")
            } catch (e: Exception) {
                onResult(false, e.message ?: "දෝෂයකි")
            }
        }
    }

    fun payRegistrationFee(onResult: (Boolean, String) -> Unit) {
        viewModelScope.launch {
            try {
                val userId = auth.currentUser?.uid ?: return@launch
                val currentFee = _uiState.value.registrationFee

                firestore.collection("users").document(userId).update("isVerified", true).await()

                val txId = java.util.UUID.randomUUID().toString()
                val tx = hashMapOf(
                    "id" to txId,
                    "userId" to userId,
                    "amount" to currentFee,
                    "type" to "FEE",
                    "description" to "Client Registration Fee",
                    "timestamp" to System.currentTimeMillis()
                )
                firestore.collection("transactions").document(txId).set(tx).await()

                onResult(true, "ගෙවීම සාර්ථකයි! ඔබගේ ගිණුම සම්පූර්ණයෙන්ම සක්‍රීය විය.")
            } catch (e: Exception) {
                onResult(false, e.localizedMessage ?: "දෝෂයකි. නැවත උත්සාහ කරන්න.")
            }
        }
    }

    fun updateProfileFull(name: String, phone: String, hourlyRate: Double, profileImage: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSaving = true, isSaveSuccess = false)
            try {
                val userId = auth.currentUser?.uid ?: return@launch
                val updates = hashMapOf<String, Any>(
                    "name" to name,
                    "phone" to phone,
                    "hourlyRate" to hourlyRate,
                    "profileImageUrl" to profileImage
                )
                firestore.collection("users").document(userId).update(updates).await()
                _uiState.value = _uiState.value.copy(isSaving = false, isSaveSuccess = true)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isSaving = false, errorMessage = e.localizedMessage)
            }
        }
    }

    fun resetSaveSuccess() { _uiState.value = _uiState.value.copy(isSaveSuccess = false) }
}

File: flutter_app/lib/presentation/profile/profile_view_model.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart' as auth;
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:uuid/uuid.dart';

import '../../domain/model/data_models.dart';

class ProfileUiState {
  final bool isLoading;
  final User? userData;
  final String? errorMessage;
  final bool isSaving;
  final bool isSaveSuccess;
  final double registrationFee;

  ProfileUiState({
    this.isLoading = true,
    this.userData,
    this.errorMessage,
    this.isSaving = false,
    this.isSaveSuccess = false,
    this.registrationFee = 1000.0,
  });

  ProfileUiState copyWith({
    bool? isLoading,
    User? userData,
    String? errorMessage,
    bool? isSaving,
    bool? isSaveSuccess,
    double? registrationFee,
  }) {
    return ProfileUiState(
      isLoading: isLoading ?? this.isLoading,
      userData: userData ?? this.userData,
      errorMessage: errorMessage ?? this.errorMessage,
      isSaving: isSaving ?? this.isSaving,
      isSaveSuccess: isSaveSuccess ?? this.isSaveSuccess,
      registrationFee: registrationFee ?? this.registrationFee,
    );
  }
}

class ProfileViewModel extends ChangeNotifier {
  final auth.FirebaseAuth _auth = auth.FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  ProfileUiState _uiState = ProfileUiState();
  ProfileUiState get uiState => _uiState;

  ProfileViewModel() {
    fetchUserProfile();
    fetchSystemSettings();
  }

  void _setUiState(ProfileUiState state) {
    _uiState = state;
    notifyListeners();
  }

  void fetchUserProfile() {
    final userId = _auth.currentUser?.uid;
    if (userId == null) return;
    
    _firestore.collection("users").doc(userId).snapshots().listen((snapshot) async {
      if (snapshot.exists) {
        final user = User.fromMap(snapshot.data()!);
        _setUiState(_uiState.copyWith(isLoading: false, userData: user));

        final isCaregiverRole = user.role == "CAREGIVER" || user.role == "NURSE";
        if (isCaregiverRole && user.kycStatus == "APPROVED" && !user.isVerified) {
          await _firestore.collection("users").doc(userId).update({"isVerified": true});
        }
      }
    }, onError: (e) {
      _setUiState(_uiState.copyWith(isLoading: false, errorMessage: e.toString()));
    });
  }

  void fetchSystemSettings() {
    _firestore.collection("settings").doc("general").snapshots().listen((snapshot) {
      if (snapshot.exists) {
        final fee = snapshot.data()?['registrationFee']?.toDouble() ?? 1000.0;
        _setUiState(_uiState.copyWith(registrationFee: fee));
      }
    });
  }

  Future<void> requestLeave(int dateMs, String reason, Function(bool, String) onResult) async {
    try {
      final userId = _auth.currentUser?.uid;
      if (userId == null) return;

      final userName = _uiState.userData?.name ?? "Caregiver";
      final dateStr = DateTime.fromMillisecondsSinceEpoch(dateMs).toIso8601String().split('T')[0];

      final leaveId = const Uuid().v4();
      final leaveReq = LeaveRequest(
        id: leaveId,
        caregiverId: userId,
        caregiverName: userName,
        dateMs: dateMs,
        dateString: dateStr,
        reason: reason,
        status: "PENDING",
        requestedAt: DateTime.now().millisecondsSinceEpoch,
      );

      await _firestore.collection("leave_requests").doc(leaveId).set(leaveReq.toMap());
      onResult(true, "නිවාඩු ඉල්ලීම යවන ලදී. Admin අනුමැතිය ලැබෙන තෙක් රැඳී සිටින්න.");
    } catch (e) {
      onResult(false, e.toString());
    }
  }

  Future<void> payRegistrationFee(Function(bool, String) onResult) async {
    try {
      final userId = _auth.currentUser?.uid;
      if (userId == null) return;

      final currentFee = _uiState.registrationFee;

      await _firestore.collection("users").doc(userId).update({"isVerified": true});

      final txId = const Uuid().v4();
      await _firestore.collection("transactions").doc(txId).set({
        "id": txId,
        "userId": userId,
        "amount": currentFee,
        "type": "FEE",
        "description": "Client Registration Fee",
        "timestamp": DateTime.now().millisecondsSinceEpoch,
      });

      onResult(true, "ගෙවීම සාර්ථකයි! ඔබගේ ගිණුම සම්පූර්ණයෙන්ම සක්‍රීය විය.");
    } catch (e) {
      onResult(false, e.toString());
    }
  }

  Future<void> updateProfileFull(String name, String phone, double hourlyRate, String profileImageBase64) async {
    _setUiState(_uiState.copyWith(isSaving: true, isSaveSuccess: false));
    try {
      final userId = _auth.currentUser?.uid;
      if (userId == null) return;

      final Map<String, dynamic> updates = {
        "name": name,
        "phone": phone,
        "hourlyRate": hourlyRate,
      };
      
      if (profileImageBase64.isNotEmpty) {
        updates["profileImageUrl"] = profileImageBase64;
      }

      await _firestore.collection("users").doc(userId).update(updates);
      _setUiState(_uiState.copyWith(isSaving: false, isSaveSuccess: true));
    } catch (e) {
      _setUiState(_uiState.copyWith(isSaving: false, errorMessage: e.toString()));
    }
  }

  void resetSaveSuccess() {
    _setUiState(_uiState.copyWith(isSaveSuccess: false));
  }
}
